### ISSUE TITLE
Evaluation script: Different results when run on 1 and on 4 devices

### ISSUE THREAD

**Issue by mseeger** [awslabs/keys_values#4]
**Is your feature request related to a problem? Please describe.**

The evaluation script `finetune/longcontext_eval.py` allows to run evaluations for several checkpoints. It supports parallelization across several devices (they do not need to communicate).

When running the script (1) on a single device and (2) on 4 devices, I am getting different loss values for the same batches. Nothing is random about evaluations (evaluation loss).

The script cannot be trusted at this point.

**Comment by mseeger** [comment]
After major refactoring and simplification of eval workflow and data iterators, I still get different results:

```
For testing, we create copies of checkpoints, so we can run independently,
using different variants: variant0_copy1, variant0_copy2.
Both have final, step-000010, step-000020

[4 devices]

CUDA_VISIBLE_DEVICES="4,5,6,7" PYTORCH_ALLOC_CONF=expandable_segments:True KEYSVALS_LOG_DIR="/home/ubuntu/out/finetune/ml_ws/lora/qwen2_5_0_5b/variant0_copy1/eval_logs"; python3 keys_values/__main__.py eval_long /home/ubuntu/out/finetune/ml_ws/lora/qwen2_5_0_5b/variant0_copy1 --model_type lora --devices 4 --batch_size 4 --kv_cache.name h2o-torch-quantized8 --kv_cache.cache_length 32768 --kv_cache.chunk_size 1024 --verbose some --attention_forward_temp_size_gb 8

[running... DONE]

[1 device]

CUDA_VISIBLE_DEVICES="0" PYTORCH_ALLOC_CONF=expandable_segments:True KEYSVALS_LOG_DIR="/home/ubuntu/out/finetune/ml_ws/lora/qwen2_5_0_5b/variant0_copy2/eval_logs0"; python3 keys_values/__main__.py eval_long /home/ubuntu/out/finetune/ml_ws/lora/qwen2_5_0_5b/variant0_copy2 --model_type lora --devices 1 --batch_size 4 --kv_cache.name h2o-torch-quantized8 --kv_cache.cache_length 32768 --kv_cache.chunk_size 1024 --verbose some --attention_forward_temp_size_gb 8
CUDA_VISIBLE_DEVICES="1" PYTORCH_ALLOC_CONF=expandable_segments:True KEYSVALS_LOG_DIR="/home/ubuntu/out/finetune/ml_ws/lora/qwen2_5_0_5b/variant0_copy2/eval_logs1"; python3 keys_values/__main__.py eval_long /home/ubuntu/out/finetune/ml_ws/lora/qwen2_5_0_5b/variant0_copy2 --model_type lora --devices 1 --batch_size 4 --kv_cache.name h2o-torch-quantized8 --kv_cache.cache_length 32768 --kv_cache.chunk_size 1024 --verbose some --attention_forward_temp_size_gb 8
CUDA_VISIBLE_DEVICES="2" PYTORCH_ALLOC_CONF=expandable_segments:True KEYSVALS_LOG_DIR="/home/ubuntu/out/finetune/ml_ws/lora/qwen2_5_0_5b/variant0_copy2/eval_logs2"; python3 keys_values/__main__.py eval_long /home/ubuntu/out/finetune/ml_ws/lora/qwen2_5_0_5b/variant0_copy2 --model_type lora --devices 1 --batch_size 4 --kv_cache.name h2o-torch-quantized8 --kv_cache.cache_length 32768 --kv_cache.chunk_size 1024 --verbose some --attention_forward_temp_size_gb 8

[0, 1 running ... STOPPED]

After fix:
[2 running ... STOPPED]

Comparing metric values:

* variant0_copy1:

idx,task,loss
0,final,1.2890625
1,final,2.125
2,final,1.5390625
3,final,3.40625
4,final,1.375
5,final,0.390625
6,final,4.78125
7,final,1.8125
8,final,3.28125
9,final,0.4765625
10,final,0.421875
11,final,1.3125
12,final,2.8125
13,final,4.25
14,final,0.443359375
15,final,5.40625
16,final,3.96875
17,final,0.462890625
18,final,4.03125
19,final,0.4453125
[...]
172,final,1.6953125
173,final,3.3125

* variant0_copy2:

idx,task,loss
0,final,1.34375
1,final,2.46875
2,final,1.40625
3,final,3.125
4,final,1.390625
5,final,0.392578125
6,final,4.59375
7,final,1.921875
8,final,3.15625
9,final,0.482421875
10,final,0.455078125
11,final,1.234375
12,final,2.796875
13,final,3.921875
14,final,0.427734375
15,final,5.625
16,final,3.265625
17,final,0.515625
18,final,3.890625
19,final,1.1484375
[...]
172,final,3.0625
173,final,3.90625
```

Different values again. Need to store and compare the exact inputs (also model weights) into the eval call.

**Comment by mseeger** [comment]
Size statistics:
```
[rank 0 | 2026-02-26 08:02:55]: Batch final, [0, 1, 2, 3]: loss = 2.094, eval_time = 208916.78 ms
Storing to /home/ubuntu/out/finetune/ml_ws/lora/qwen2_5_0_5b/variant0_copy2/final/eval/eval_metrics_rest0.csv
DEBUG: Loading A state from /home/ubuntu/out/finetune/ml_ws/lora/qwen2_5_0_5b/variant0_copy2/final/eval/eval_metrics_rest0.pth
DEBUG: Comparing A state against current state (B):
Comparing data_batch dictionaries:
    input_ids
    targets
Comparing gpt_state_dict dictionaries:
    lm_head.linear.weight:                   0.01:1.85e-04|0.50:1.01e-02|0.90:2.55e-02|0.99:4.20e-02
    transformer.h.0.attn.proj.linear.weight: 0.01:1.30e-04|0.50:7.39e-03|0.90:2.10e-02|0.99:4.20e-02
    transformer.h.0.attn.qkv.linear.bias:    0.01:1.22e-03|0.50:1.27e+00|0.90:1.12e+01|0.99:6.56e+01
    transformer.h.0.attn.qkv.linear.weight:  0.01:3.45e-04|0.50:2.33e-02|0.90:9.18e-02|0.99:2.45e-01
    transformer.h.0.attn.qkv.lora_A:         0.01:3.34e-04|0.50:1.72e-02|0.90:3.66e-02|0.99:4.98e-02
    transformer.h.0.attn.qkv.lora_B:         0.01:1.91e-04|0.50:8.12e-03|0.90:1.86e-02|0.99:2.60e-02
    transformer.h.0.mlp.fc_1.linear.weight:  0.01:2.71e-04|0.50:1.48e-02|0.90:3.78e-02|0.99:6.59e-02
    transformer.h.0.mlp.fc_2.linear.weight:  0.01:2.24e-04|0.50:1.23e-02|0.90:3.05e-02|0.99:4.88e-02
    transformer.h.0.mlp.proj.linear.weight:  0.01:2.15e-04|0.50:1.17e-02|0.90:3.00e-02|0.99:5.08e-02
    transformer.h.0.norm_1.weight:           0.01:2.32e-02|0.50:4.64e-02|0.90:1.22e-01|0.99:3.79e-01
    transformer.h.0.norm_2.weight:           0.01:5.62e-01|0.50:6.37e-01|0.90:7.60e-01|0.99:1.50e+00

    transformer.h.1.attn.proj.linear.weight: 0.01:1.55e-04|0.50:8.97e-03|0.90:2.48e-02|0.99:4.39e-02
    transformer.h.1.attn.qkv.linear.bias:    0.01:1.43e-03|0.50:3.03e-01|0.90:1.69e+00|0.99:8.71e+00
    transformer.h.1.attn.qkv.linear.weight:  0.01:2.59e-04|0.50:1.54e-02|0.90:4.81e-02|0.99:9.13e-02
    transformer.h.1.attn.qkv.lora_A:         0.01:3.99e-04|0.50:1.77e-02|0.90:3.93e-02|0.99:5.49e-02
    transformer.h.1.attn.qkv.lora_B:         0.01:1.27e-04|0.50:7.78e-03|0.90:1.75e-02|0.99:2.60e-02
    transformer.h.1.mlp.fc_1.linear.weight:  0.01:2.78e-04|0.50:1.52e-02|0.90:3.83e-02|0.99:6.45e-02
    transformer.h.1.mlp.fc_2.linear.weight:  0.01:2.11e-04|0.50:1.15e-02|0.90:2.86e-02|0.99:4.59e-02
    transformer.h.1.mlp.proj.linear.weight:  0.01:2.05e-04|0.50:1.12e-02|0.90:2.86e-02|0.99:4.64e-02
    transformer.h.1.norm_1.weight:           0.01:1.57e-01|0.50:6.39e-01|0.90:9.32e-01|0.99:2.24e+00
    transformer.h.1.norm_2.weight:           0.01:3.69e-01|0.50:1.05e+00|0.90:1.25e+00|0.99:1.33e+00
    [...]
    transformer.ln_f.weight:                 0.01:5.31e+00|0.50:6.78e+00|0.90:7.66e+00|0.99:1.25e+01
    transformer.wte.weight:                  0.01:1.85e-04|0.50:1.01e-02|0.90:2.55e-02|0.99:4.20e-02
Comparing head_state_dict dictionaries:
Comparing intermediates dictionaries:
    forward_wte_0:32768
    forward_wte_32768:41984
    forward_wte_41984:51200
    forward_wte_51200:60416
    forward_wte_60416:69632
    forward_wte_69632:78848
    forward_wte_78848:88064
    forward_wte_88064:97280
    forward_wte_97280:106496
    forward_wte_106496:115712
    forward_wte_115712:124928
    forward_wte_124928:134144
    forward_wte_134144:143360
    forward_wte_143360:152576
    forward_wte_152576:154243
    forward_block0_0:32768_0:32768
Significant differences A vs B for forward_block0_0:32768_0:32768
Caught exception during evaluation:
Tensor-likes are not close!

Mismatched elements: 85527529 / 117440512 (72.8%)
Greatest absolute difference: 1.359375 at index (2, 5344, 62) (up to 1e-05 allowed)
Greatest relative difference: inf at index (0, 19, 457) (up to 0.016 allowed)
```

These weight values are not very large or very small. `lora_B`, `lora_A` are rather small. `attn.qkv.linear.bias` are the largest, but only up to 65. This cannot explain the differences!

**Comment by mseeger** [comment]
Another issue with the eval script: File-locking does often not work, in that two devices pick the same batch!
```
eval_logs/gpu0.log:[rank 0 | 2026-02-27 08:31:35]: Running inference for batch final, [8, 9, 10, 11]
eval_logs/gpu1.log:[rank 1 | 2026-02-27 08:31:35]: Running inference for batch final, [0, 1, 2, 3]
eval_logs/gpu2.log:[rank 2 | 2026-02-27 08:31:35]: Running inference for batch final, [4, 5, 6, 7]
eval_logs/gpu3.log:[rank 3 | 2026-02-27 08:31:35]: Running inference for batch final, [4, 5, 6, 7]
```

==> Fix this as well!
